network/p2p/handler_test.go

```diff
@@ -112,3 +116,166 @@ func TestValidatorHandlerAppRequest(t *testing.T) {
 		})
 	}
 }
+
+func TestNewDynamicThrottlerHandler(t *testing.T) {
+	type request struct {
+		nodeID  ids.NodeID
+		wantErr *common.AppError
+	}
+
+	tests := []struct {
+		name             string
+		throttlingPeriod time.Duration
+		requestsPerPeer  int
+
+		// must be same length
+		validatorSets [][]ids.NodeID
+		requests      [][]request
+	}{
+		{
+			name: "validator request",
+			validatorSets: [][]ids.NodeID{
+				{
+					{1},
+				},
+			},
+			requests: [][]request{
+				{
+					{nodeID: ids.NodeID{1}},
+				},
+			},
+		},
+		{
+			name:             "validator requests not throttled",
+			throttlingPeriod: time.Hour,
+			requestsPerPeer:  5,
+			validatorSets: [][]ids.NodeID{
+				{
+					{1},
+				},
+			},
+			requests: [][]request{
+				{
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}},
+				},
+			},
+		},
+		{
+			name:             "validator request throttled",
+			throttlingPeriod: time.Hour,
+			requestsPerPeer:  3,
+			validatorSets: [][]ids.NodeID{
+				{
+					{1},
+				},
+			},
+			requests: [][]request{
+				{
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}, wantErr: ErrThrottled},
+				},
+			},
+		},
+		{
+			name:             "validator request throttled before update",
+			throttlingPeriod: time.Hour,
+			requestsPerPeer:  3,
+			validatorSets: [][]ids.NodeID{
+				{
+					{1},
+				},
+				{
+					{1}, {2}, {3}, {4}, {5}, {6}, {7}, {8}, {9},
+				},
+			},
+			requests: [][]request{
+				{
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}},
+				},
+				{
+					{nodeID: ids.NodeID{1}, wantErr: ErrThrottled},
+				},
+			},
+		},
+		{
+			name:             "validator request throttled after update",
+			throttlingPeriod: time.Hour,
+			requestsPerPeer:  3,
+			validatorSets: [][]ids.NodeID{
+				{
+					{1},
+				},
+				{
+					{1},
+					{2},
+					{3},
+					{4},
+					{5},
+					{6},
+					{7},
+					{8},
+					{9},
+					{10},
+					{11},
+					{12},
+					{13},
+					{14},
+					{15},
+					{16},
+					{17},
+					{18},
+					{19},
+					{20},
+				},
+			},
+			requests: [][]request{
+				{
+					{nodeID: ids.NodeID{1}},
+				},
+				{
+					{nodeID: ids.NodeID{1}},
+					{nodeID: ids.NodeID{1}, wantErr: ErrThrottled},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+
+			if len(tt.validatorSets) != len(tt.requests) {
+				require.Fail("test vectors must be of same length")
+			}
```

Comment:
```suggestion
			require.Equal(len(tt.validatorSets), len(tt.requests))
```
is a one liner better here?

Reply:
I can use `require.Len()`... not sure what I was doing here...